test/rules/use_build_context_synchronously.dart

```diff
@@ -0,0 +1,269 @@
+  }
+
+  // Method given a build context to use.
+  void methodWithBuildContextParameter1(BuildContext context) async {
+    Navigator.of(context).pushNamed('routeName'); // OK
+
+    await Future<void>.delayed(Duration());
+    Navigator.of(context).pushNamed('routeName'); // LINT
+  }
+
+  // Same as above, but using a conditional path.
+  void methodWithBuildContextParameter2(BuildContext context) async {
+    if (defaultTargetPlatform == TargetPlatform.iOS) {
+      await Future<void>.delayed(Duration());
+    }
+    Navigator.of(context).pushNamed('routeName'); // LINT
+  }
+
+  // Another conditional path.
+  void methodWithBuildContextParameter2a(BuildContext context) async {
+    bool f() => true;
+    while (f()) {
+      await Future<void>.delayed(Duration());
+    }
+    Navigator.of(context).pushNamed('routeName'); // LINT
+  }
+
+  // And another.
+  void methodWithBuildContextParameter2b(BuildContext context) async {
+    for (var i = 0; i < 1; ++i) {
+      await Future<void>.delayed(Duration());
+    }
+    Navigator.of(context).pushNamed('routeName'); // LINT
+  }
+
+  // And another.
+  void methodWithBuildContextParameter2c(BuildContext context) async {
+    for (var i in [1]) {
+      await Future<void>.delayed(Duration());
+    }
+    Navigator.of(context).pushNamed('routeName'); // LINT
+  }
+
+  // And another.
+  void methodWithBuildContextParameter2d(BuildContext context) async {
+    bool f() => true;
+    do {
+      await Future<void>.delayed(Duration());
+    } while (f());
+    Navigator.of(context).pushNamed('routeName'); // LINT
+  }
+
+  Future<void> methodWithBuildContextParameter2e(BuildContext context) async {
+    await Future<void>.delayed(Duration());
+    if (!mounted) return;
+    unawaited(methodWithBuildContextParameter2e(context)); //OK
+  }
+
+  // Mounted checks are deliberately naive.
+  void methodWithBuildContextParameter3(BuildContext context) async {
+    Navigator.of(context).pushNamed('routeName'); // OK
+
+    await Future<void>.delayed(Duration());
+
+    if (!mounted) return;
+
+    // Mounted doesn't cover provided context but that's by design.
+    Navigator.of(context).pushNamed('routeName'); // OK
+  }
+
+  void methodWithMountedFieldCheck(
+      BuildContext context, WidgetStateContext stateContext) async {
+    await Future<void>.delayed(Duration());
+    if (!stateContext.mounted) return;
+    f(context); // OK
+  }
+
+  @override
+  Widget build(BuildContext context) => const Placeholder();
+}
+
+void topLevel(BuildContext context) async {
+  Navigator.of(context).pushNamed('routeName'); // OK
+
+  await Future<void>.delayed(Duration());
+  Navigator.of(context).pushNamed('routeName'); // LINT
+}
+
+void topLevel2(BuildContext context) async {
+  Navigator.of(context).pushNamed('routeName'); // OK
+
+  await Future<void>.delayed(Duration());
+  // todo (pq): validate other conditionals (for, while, do, ...)
+  // OR: should that be disallowed in another lint?
```

Comment:
Even if it's disallowed in another lint we won't know that both lints are always enabled together, so we need to decide what the right behavior is without depending on another lint.